core/src/main/java/org/apache/druid/java/util/http/client/pool/ResourcePool.java

```diff
@@ -212,15 +259,21 @@ V get()
         if (closed) {
           log.info(StringUtils.format("get() called even though I'm closed. key[%s]", key));
           return null;
-        } else if (!resourceHolderList.isEmpty()) {
-          ResourceHolder<V> holder = resourceHolderList.removeFirst();
-          if (System.currentTimeMillis() - holder.getLastAccessedTime() > unusedResourceTimeoutMillis) {
-            factory.close(holder.getResource());
+        } else if (numLentResources + deficit < maxSize) {
+          // Attempt to take an existing resource or create one if list is empty
+          if (resourceHolderList.isEmpty()) {
             poolVal = factory.generate(key);
           } else {
-            poolVal = holder.getResource();
+            ResourceHolder<V> holder = resourceHolderList.removeFirst();
+            if (System.currentTimeMillis() - holder.getLastAccessedTime() > unusedResourceTimeoutMillis) {
+              factory.close(holder.getResource());
+              poolVal = factory.generate(key);
+            } else {
+              poolVal = holder.getResource();
+            }
           }
         } else if (deficit > 0) {
+          // We are allowed to generate to fill the deficit objects
```

Comment:
what are deficit objects? From your earlier comment, deficit is just the number of exceptions.

Reply:
While maxNum is the limit for number of resources that can be allocated at any point in time, it is done with the aid of both numLentResources and deficit.

As long as numLentResources + deficit < maxNum:
   Try to take a resource from the existing pool or create one
Else if deficit > 0:
    Decrement deficit and attempt to create a new resource which will be then returned to resourceHolderList